### PR TITLE
New compression level option support

### PR DESCRIPTION
**Pull request by palant**
## Description
This introduces a new `--compression-level=fastest|best|default` setting with `default` being the default. While `default` is meant to be algorithm-specific, currently level 4 seems to be the perfect sweet spot for all supported algorithms (see benchmarking results below).

This increases the data size somewhat: the 144 KiB `large.txt` file in our tests used to compress into 17.86 KiB with gzip (level 6), new default is 21.90 KiB. Brotli used to produce 14.00 KiB, now it is 21.51 KiB. It also improves the performance significantly however, especially for Brotli that for some reason goes with “best but horribly slow compression” as default.

## Related Issue
#380

## How Has This Been Tested?
Functional testing on Linux looks good.

### CODE REVIEW

**Comment by palant** [comment]
Having played around with this a bit, it doesn’t look like the provided settings are useful enough. `fastest` is good for performance but produces really poor file sizes. And `default` is absolutely random: doing okay’ish for gzip but going to level 11 for Brotli which then takes seconds for a single file.

I think we need to redefine `default` to some meaningful tradeoff between performance and file size for all algorithms, then actually have it used by default.

**Comment by palant** [comment]
I did some local benchmarking with our `large.txt` (only the pure compression functionality, without the web server around it). Based on these numbers I’d say that the sweet spot for all ( :open_mouth: ) algorithms is level 4. That’s where the file size is already fairly close to optimal yet the performance didn’t degrade too much. Guess that’s what our default should be.

### gzip

| Level | Size | Time |
|---|---|---|
| 1 | 52.03KiB | 1.55ms |
| 2 | 34.98KiB | 1.59ms |
| 3 | 21.29KiB | 2.07ms |
| 4 | 21.90KiB | 2.13ms |
| 5 | 19.58KiB | 2.60ms |
| 6 | 17.86KiB | 4.02ms |
| 7 | 17.83KiB | 4.22ms |
| 8 | 17.82KiB | 4.40ms |
| 9 | 17.82KiB | 4.53ms |

### deflate

| Level | Size | Time |
|---|---|---|
| 1 | 52.01KiB | 1.37ms |
| 2 | 34.96KiB | 1.59ms |
| 3 | 21.27KiB | 2.04ms |
| 4 | 21.88KiB | 2.10ms |
| 5 | 19.57KiB | 2.81ms |
| 6 | 17.84KiB | 3.97ms |
| 7 | 17.81KiB | 4.65ms |
| 8 | 17.80KiB | 4.76ms |
| 9 | 17.80KiB | 4.44ms |

### brotli

| Level | Size | Time |
|---|---|---|
| 0 | 33.52KiB | 0.71ms |
| 1 | 27.82KiB | 0.95ms |
| 2 | 28.09KiB | 2.06ms |
| 3 | 24.71KiB | 2.07ms |
| 4 | 21.51KiB | 2.19ms |
| 5 | 17.53KiB | 3.31ms |
| 6 | 16.10KiB | 3.82ms |
| 7 | 15.53KiB | 12.79ms |
| 8 | 15.33KiB | 6.00ms |
| 9 | 15.27KiB | 9.68ms |
| 10 | 14.34KiB | 103.15ms |
| 11 | 14.00KiB | 283.30ms |

### zstd

| Level | Size | Time |
|---|---|---|
| 1 | 24.99KiB | 0.45ms |
| 2 | 27.32KiB | 0.49ms |
| 3 | 22.82KiB | 0.48ms |
| 4 | 22.78KiB | 0.58ms |
| 5 | 19.77KiB | 1.35ms |
| 6 | 18.13KiB | 1.76ms |
| 7 | 16.88KiB | 1.98ms |
| 8 | 16.39KiB | 2.40ms |
| 9 | 16.38KiB | 2.71ms |
| 10 | 15.89KiB | 3.86ms |
| 11 | 15.69KiB | 4.25ms |
| 12 | 15.69KiB | 24.83ms |
| 13 | 15.84KiB | 26.19ms |
| 14 | 15.69KiB | 38.09ms |
| 15 | 15.67KiB | 50.01ms |
| 16 | 14.94KiB | 44.54ms |
| 17 | 14.88KiB | 56.08ms |
| 18 | 14.79KiB | 76.51ms |
| 19 | 14.67KiB | 144.52ms |
| 20 | 14.67KiB | 186.95ms |
| 21 | 14.67KiB | 279.17ms |
| 22 | 14.67KiB | 466.26ms |

**Comment by palant** [comment]
Looking at other web servers for comparison, nginx by default produces gzip sizes around level 2, Apache between level 5 and 6, caddy  around level 3 and 4, lighttpd level 6.

File: docs/content/configuration/environment-variables.md [hunk 1]
```diff
@@ -80,6 +80,9 @@ Specify an optional CORS list of exposed HTTP headers separated by commas. It re
 ### SERVER_COMPRESSION
 `Gzip`, `Deflate` or `Brotli` compression on demand determined by the `Accept-Encoding` header and applied to text-based web file types only. See [ad-hoc mime-type list](https://github.com/static-web-server/static-web-server/blob/master/src/compression.rs#L20). Default `true` (enabled).
```
Comment (gaby) [inline docs/content/configuration/environment-variables.md]: This is missing mention of `zstd`

File: src/compression.rs [hunk 2]
```diff
@@ -268,10 +288,17 @@ pub fn brotli(
 pub fn zstd(
     mut head: http::response::Parts,
     body: CompressableBody<Body, hyper::Error>,
+    level: CompressionLevel,
 ) -> Response<Body> {
+    const DEFAULT_COMPRESSION_LEVEL: i32 = 4;
```
Comment (gaby) [inline src/compression.rs]: This should be `3`, which is the default zstd compression used everywhere.

File: docs/content/configuration/environment-variables.md [hunk 3]
```diff
@@ -80,6 +80,9 @@ Specify an optional CORS list of exposed HTTP headers separated by commas. It re
...
 
+### SERVER_COMPRESSION_LEVEL
+Supported values are `fastest` (fast compression but larger resulting files), `best` (smallest file size but potentially slow) and `default` (algorithm-specific balanced compression level). Default is `default`.
+
 ### SERVER_COMPRESSION_STATIC
 Look up the pre-compressed file variant (`.gz` or `.br`) on disk of a requested file and serves it directly if available. Default `false` (disabled). The compression type is determined by the `Accept-Encoding` header.
```
Comment (gaby) [inline docs/content/configuration/environment-variables.md]: This is missing mention of `.zst`

**Comment by gaby** [comment]
@palant I sent a PR to the `benchmark-setup`  repo to enable zstd and brotli in Caddyfile.

**Review by gaby — APPROVED** [review]

File: docs/content/configuration/config-file.md [hunk 4]
```diff
@@ -28,6 +28,7 @@ cache-control-headers = true
 
 #### Auto Compression
 compression = true
+compression_level = "default"
```
Comment (joseluisq) [inline docs/content/configuration/config-file.md]:
```suggestion
compression-level = "default"
```

**Review by joseluisq — CHANGES_REQUESTED** [review]
Looks good to me. Just minor change and maybe check if we can make some new APIs internal.

**Comment by palant** [comment]
> and maybe check if we can make some new APIs internal.

Ok, I made `CompressionLevel::into_algorithm_level()` private. The only other new thing is `CompressionLevel` itself, and it has to be public because it is used in already public structs.

**Review by joseluisq — CHANGES_REQUESTED** [review]
Last thing, please update the compression page and mention the new compression-level feature.
https://github.com/static-web-server/static-web-server/blob/master/docs/content/features/compression.md

**Comment by palant** [comment]
> Last thing, please update the compression page and mention the new compression-level feature. https://github.com/static-web-server/static-web-server/blob/master/docs/content/features/compression.md

That text generally isn’t very up-to-date, so I had to rewrite it quite significantly.

**Review by joseluisq — APPROVED** [review]
Thanks!

**Pull request merged**